content/posts/2018-01-10-kotlin-null.md

```diff
@@ -0,0 +1,249 @@
+_will_ make you forget about simple things — like the one we are discussing here — Java does not have
+nullability metadata out of the box. You will expect that everything you touch actually has it
+(a classic [Midas mistake](https://en.wikipedia.org/wiki/Midas#Golden_Touch)),
+because your shiny Kotlin codebase has it and you have a great confidence in that.
+
+Prepare for unforeseen consequences.
+
+## The Solution
+
+### The Obvious
+
+This is a hard one since it is based on a non-automated semantics mechanism — the mind.
+It is always better to remember when you are using Java code and when you are using the Kotlin one.
+
+Ask more _strange_ questions during a code review — _Are we guarded against this Java code?_
+It may sound dumb and people are often too lazy to check but it is better to raise a question instead of doing nothing.
+
+### Contributions
+
+Let’s face it — a lot of code we use as libraries is open source.
+The entire Android Framework is here — clone it, read it, modify it.
+Yes, it is not so easy to talk people into using nullability annotations,
+especially if your main argument is “I use Kotlin, so you should...”.
+
+* Maintainers are not your personal minions, so please show at least some respect.
+* The Kotlin argument does not work all the time and it should be the last one you have up your sleeve.
+  Not so many people actually use Kotlin at this point, but there is a huge amount of Java code.
+* Nullability annotations improve documentation and IDE support for everyone who uses the Java code,
+  so it seems like a better argument.
+* Pull Requests are always welcome!
+
+### Facade the Hell of It
+
+The worst case scenario — you do not have access to the source code, maintainers are quiet for years.
+
+Make your own abstraction using Kotlin! Believe me, almost everything is doable with enough effort
+(excluding codebases with a huge number of public methods — like RxJava).
+As a side benefit — you will make your code more testable encapsulating implementation behind an interface.
+
+Sample time!
+
+```java
+class CorporateCalculator {
+    void calculate(Argument argument) {
+        if (argument == null) throw new NullPointerException("Gotcha!");
+
+        calculateUsingCorporateBlackMagic(argument);
+    }
+}
+```
+```kotlin
+interface Calculator {
+    fun calculate(argument: Argument)
+
+    class CorporateImpl : Calculator {
+        override fun calculate(argument: Argument) {
+            CorporateCalculator().calculate(argument)
+        }
+    }
+}
+```
+
+As you see, we hide the `CorporateCalculator` behind a Kotlin-based abstraction
+which will save us from passing `null` to the `calculate` method.
+
+# Making a Swift Turn
+
+The good news is — we are not alone. Like, on the planet!
+
+The Java and Kotlin situation is kind of similar to the Objective-C and Swift one.
+Objective-C objects are essentially C pointers, which can be `NULL` or `nil` in Objective-C terms.
```

Comment:
Object is not a pointer though, I guess you can just say "Objective-C pointers can be NULL or nil"

Reply:
I think technically @ming13  is correct  but @artem-zinnatullin suggestion reads better imho.
I would also use just `nil` or just `null` but not both as for somebody who hasn't done iOS development it may sounds like a A or B choice. Alternative, a compromise could be adding a clarification note in parenthesis like:

> which can be `nil` (the objective-c counterpart to `null`)

If you think that this is duplicate because you have already mentioned it somewhere above, then I would argue only use `nil`.

Reply:
Don’t forget about `Nil` and `NSNull`!

Thank you guys, I’ll change this to exclude pointers and use `nil`-only to avoid a possible confusion.